tv_apps/tv-deck/js/channel_manager.js

```diff
@@ -25,7 +30,305 @@
+   * We have to scan the channels before calling getChannels.
+   */
+  proto.scanChannels = function cm_scanChannels() {
+    this.isScanning = true;
+    var source = this.playingSource.source;
+    this.playingSource.channels = [];
+    this.playingSource.channelIndexHash = {};
+    if (source.isScanning) {
+      return;
+    }
+
+    var onscanningstatechanged = function(event) {
+      var state = event.state;
+      switch (state) {
+        case 'completed':
+        /* falls through */
+        case 'stopped':
+          this._onScanningCompleted();
+          source.removeEventListener(
+                                'scanningstatechanged', onscanningstatechanged);
+          break;
+        default:
+          break;
+      }
+    }.bind(this);
+
+    source.addEventListener('scanningstatechanged', onscanningstatechanged);
+
+    source.startScanning({
+      isRescanned: true
+    });
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * Sort channels for channel switching.
+   */
+  proto._onScanningCompleted = function cm_onScanningCompleted() {
+    this.playingSource.source.getChannels().then(function onsuccess(channels) {
+      this.isScanning = false;
+      if (channels.length === 0) {
+        console.log('Error, no channel found!');
+        return;
+      }
+
+      // Sort channels. Channel number can be XX-XX-XX
+      channels.sort(function(channelA, channelB) {
+        return this._compareChannel(channelA.number, channelB.number);
+      }.bind(this));
+
+      var i;
+      for (i = 0; i < channels.length; i++) {
+        this.playingSource.channelIndexHash[channels[i].number] = i;
+        this.playingSource.channels[i] = {
+          channel: channels[i],
+          programs: {},
+        };
+      }
+
+      if (!this.playingChannelNumber) {
+        this.playingChannelNumber = channels[0].number;
+      }
+
+      if (!this.playingChannel) {
+        this.fire('error');
+        return;
+      }
+
+      this.fire('tune');
+    }.bind(this), function onerror(error) {
+      alert(error);
+    });
+  };
+
+  proto.setPlayingSource = function cm_setPlayingSource(callback) {
+    if (this.isScanning) {
+      return;
+    }
+
+    if (this.playingTuner.tuner.currentSource === this.playingSource.source) {
+      this.setPlayingChannel(callback);
+      return;
+    }
+
+    this.playingTuner.tuner.setCurrentSource(this.playingSourceType)
+      .then(function() {
+        this.setPlayingChannel(callback);
+      }.bind(this), function() {
+        alert('Source not found');
+      });
+  };
+
+  proto.setPlayingChannel = function cm_setPlayingChannel(callback) {
+    if (this.isScanning) {
+      return;
+    }
+
+    this.playingSource.source.setCurrentChannel(this.playingChannelNumber)
+      .then(function() {
+        if (callback) {
+          callback();
+        }
+      }.bind(this), function() {
+        alert('Channel not found');
```

Comment:
DO NOT use alert. Just use console.error. Please ask UX for the text here and put it to l10n.